File: chapter01/src/main/java/step01/Bag.java
package step01;

import lombok.Data;

@Data
public class Bag {
    private Long amount;
    private Inviation inviation;
    private Ticket ticket;

    public Bag(Long amount) {
        this(null, amount);
    }

    public Bag(Inviation inviation, long amount) {
        this.inviation = inviation;
        this.amount = amount;
    }

    public boolean hasInvitation() {
        return inviation != null;
    }

    public boolean hasTicket() {
        return ticket != null;
    }

    public void minuAmount(Long amount) {
        this.amount -= amount;
    }

    public void plusAmount(Long amount) {
        this.amount += amount;
    }
}

File: chapter06/src/main/kotlin/Bag.kt
data class Bag(var amount: Long,
               val invitation: Invitation,
               private var ticket: Ticket) {
    fun setTicket(ticket: Ticket): Long {
        return if (hasInvitation()) {
            this.ticket = ticket
            0L
        } else {
            this.ticket = ticket
            minusAmount(ticket.fee)
            ticket.fee
        }
    }

    fun hasInvitation(): Boolean {
        return true
    }

    fun minusAmount(fee: Long) {
        this.amount -= fee
    }

}
